documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,327 @@
+        /// <returns>
+        /// true if stopping the tracing session succeeded. false otherwise. 
+        /// </returns> 
+        public static GenerateCoreDump(int processId, string dumpName, DumpType dumpType, bool logDumpGeneration)
+
+        /// <summary>
+        /// Attach a profiler.
+        /// </summary>
+        /// <param name="processId">Target process' ID</param>
+        /// <param name="attachTimeout">Timeout in seconds</param>
+        /// <param name="attachTimeout">
+        public static int AttachProfiler(int processId, uint attachTimeout, Guid profilerGuid, string profilerPath, byte[] additionalData);
+
+    }
+}
+```
+
+
+### Exceptions that can be thrown 
+
+```cs
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    // Generic wrapper for exceptions thrown by this library
+    public class DiagnosticsClientExcpetion : Exception {} 
+
+    // When a certian command is not supported by either the library or the target process' runtime
+    public class UnsupportedProtocolException : DiagnosticsClientException {} 
+
+    // When the runtime is no longer availble for attaching.
+    public class ServerNotAvailableException : DiagnosticsClientException {} 
+
+    // When the runtime responded with an error
+    public class ServerThrownException : DiagnosticsClientException {} 
+}
+```
+
+## Micorosft.Diagnostics.Client.EventPipe namespace
+
+### EventPipeProvider
+A class that describes an EventPipe provider.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class EventPipeProvider
+    {
+        public EventPipeProvider(
+            string name,
+            long keywords = ulong.MaxValue,
+            EventLevel eventLevel = EventLevel.Verbose,
+            string filterData = null)
+
+        public long Keywords { get; }
+
+        public EventLevel EventLevel { get; }
+
+        public string Name { get; }
+
+        public IEnumerable<KeyValuePair<string, string>> FilterData { get; }
+
+        public override string ToString();
+        
+        public override bool Equals(object obj);
+
+        public override int GetHashCode();
+
+        public static bool operator ==(Provider left, Provider right);
+
+        public static bool operator !=(Provider left, Provider right);
+    }
+}
+```
+
+### EventPipeSession
+This is a class to represent an EventPipeSession. It is meant to be immutable and acts as a handle to each session that has been started. 
+
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class EventPipeSession
+    {
+        public EventPipeSession(
+            int processId,
+            long sessionId,
+            IEnumerable<EventPipeProvider> providers,
+            int circularBufferMB,
+            bool rundownRequested,
+            Stream stream
+        )
+
+        public int ProcessId { get; }
+        public int SessionId { get; } 
+        public IReadOnlyList<EventPipeProvider> Providers { get ; }
```

Comment:
I think we will want to add APIs for changing these values _during_ a session.  Would we not want to be able to read these values as well before we update them?  I would imagine a consumer would maybe want to do something like
```csharp
session.AddProvider(p1);
if (session.BufferSize < y)
    session.SetBufferSize(y);
Thread.Sleep(10_000);
session.RemoveProvider(p1);
session.AddProvider(p2); // Where p2 is less chatty than p1
session.SetBufferSize(z); // where z < y
```

For example, when collecting a gcdump, we have to create multiple sessions to achieve what one mutable session could.  I definitely see your point though.

Reply:
I'd try to err on the side of keeping APIs simple. If I can't point to a clear and valuable scenario that needs the API then I'd rather not add it. Correcting for too few APIs later is easy in comparison to trying to deprecate an API we regret adding.

I'd also be worried that these getters have more complexity than it may first appear - they represent a cache in a distributed system. If we assume the backing config data for a session in a future runtime version has an IPC protocol to mutate it then the mutations aren't forced to be transmitted by calls through this API wrapper. The cache could become unsynchronized because of out-of-band updates. Another challenge is that in the future we might want to add an IPC protocol command that reads the session state in addition to the one that updates it. If we did that we'd either need to convert the behavior of the cached getter API to become an uncached API that behaves a little differently, or introduce multiple similarly named APIs to handle both cached and uncached variations.